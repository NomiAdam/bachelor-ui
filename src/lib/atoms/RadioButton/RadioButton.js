import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { lightTheme } from '../../constants/theme';

const StyledInput = styled.input`
  position: absolute;
  opacity: 0;
  cursor: pointer;
`;

const StyledSpan = styled.span`
  margin: 3.5px;
  position: absolute;
  top: 0;
  left: 0;
  height: 25px;
  width: 25px;
  background-color: ${({ checked }) => checked ? lightTheme.lightBlue : lightTheme.whiteBlue};
  border-radius: 50%;
  transition: .25s all ease;
  &:after {
    content: "";
    position: absolute;
    display: ${({ checked }) => checked ? 'block' : 'none'};
	top: 9px;
	left: 9px;
	width: 8px;
	height: 8px;
	border-radius: 50%;
	background: ${lightTheme.white};
  }
`;

const StyledContainer = styled.label`
  display: block;
  position: relative;
  cursor: pointer;
  font-size: 22px;
  user-select: none;
  &:hover {
    ${StyledSpan} {
      background-color: ${lightTheme.whiteBlue};
    }
  }
`;

const StyledLabel = styled.span`
  margin-left: 35px;
  padding: 0; 
`;

const RadioButton = ({label, checked, handleChange, handleClick}) => {
    return (
        <StyledContainer>
            <StyledLabel>{label}</StyledLabel>
            <StyledInput type="radio" checked={checked} onChange={handleChange} onClick={handleClick} />
            <StyledSpan checked={checked} />
        </StyledContainer>
    );
};

RadioButton.propTypes = {
    /**
     * String value of label, label will reside on left side of radio
     */
    label: PropTypes.string,
    /**
     * Whether radio is checked or not
     */
    checked: PropTypes.bool,
    /**
     * Function to be called upon change
     */
    handleChange: PropTypes.func,
    /**
     * Function to be called upon click
     */
    handleClick: PropTypes.func,
};

export { StyledSpan };
export default RadioButton;
