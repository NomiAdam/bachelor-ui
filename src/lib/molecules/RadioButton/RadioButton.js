import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { path } from 'ramda';
import { basicTheme, lightTheme } from '../../constants/theme';
import { resolveTheme, TYPES } from '../../utils/resolveTheme';

const StyledInput = styled.input`
  position: absolute;
  opacity: 0;
`;

const StyledSpan = styled.span`
  margin: 3.5px;
  position: absolute;
  top: 0;
  left: 0;
  height: 25px;
  width: 25px;
  background-color: ${ ({ checked, ...props }) => (checked ? resolveTheme(TYPES.COLOR)(props) : basicTheme.grey) };
  border-radius: 50%;
  transition: .25s all ease;
  &:after {
    content: "";
    position: absolute;
    display: ${ ({ checked }) => (checked ? 'block' : 'none') };
	  top: 8.5px;
	  left: 8.5px;
	  width: 8px;
	  height: 8px;
	  border-radius: 50%;
	  background: ${ lightTheme.white };
  }
`;

const StyledContainer = styled.label`
  display: block;
  position: relative;
  cursor: pointer;
  font-size: 22px;
  user-select: none;
  &:hover {
    cursor: ${ ({ disabled }) => (disabled ? 'not-allowed' : 'pointer') };
    ${ StyledSpan } {
      background-color: ${ resolveTheme(TYPES.BACKGROUND_COLOR_HOVER) };
    }
  }
`;

const StyledLabel = styled.span`
  margin-left: 35px;
  padding: 0; 
`;

const getTargetValue = path(['target', 'value']);
const RadioButton = ({
    label, checked, handleChange, handleClick, disabled,
}) => (
    <StyledContainer disabled={ disabled }>
        <StyledLabel>{label}</StyledLabel>
        <StyledInput
            type="radio"
            checked={ checked }
            onChange={ disabled ? undefined : e => handleChange(e, getTargetValue(e)) }
            onClick={ handleClick }
            disabled={ disabled }
        />
        <StyledSpan checked={ checked } />
    </StyledContainer>
);

RadioButton.propTypes = {
    /**
   * Whether its disabled or not
   */
    disabled: PropTypes.bool,
    /**
     * String value of label, label will reside on left side of radio
     */
    label: PropTypes.string,
    /**
     * Whether radio is checked or not
     */
    checked: PropTypes.bool,
    /**
     * Function to be called upon change
     */
    handleChange: PropTypes.func,
    /**
     * Function to be called upon click
     */
    handleClick: PropTypes.func,
};

export { StyledSpan };
export default RadioButton;
