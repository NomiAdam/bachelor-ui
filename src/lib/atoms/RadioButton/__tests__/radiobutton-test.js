/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import RadioButton, { StyledSpan } from '../RadioButton';
import 'jest-styled-components';

describe('RadioButton', () => {
    it('should be defined', () => {
        expect(shallow(<RadioButton />)).toBeDefined();
    });
    it('should match the snapshot', () => {
        expect(shallow(<RadioButton />)).toMatchSnapshot();
    });
    it('should match the snapshot when checked', () => {
        expect(shallow(<RadioButton checked />)).toMatchSnapshot();
    });
    it('should match the snapshot with label', () => {
        expect(shallow(<RadioButton label="Mock" />)).toMatchSnapshot();
    });
    it('should have display block when checked', () => {
        const wrapper = shallow(<RadioButton checked />);
        expect(wrapper.find(StyledSpan)).toHaveStyleRule('display', 'block', {
            modifier: ':after',
        });
    });
    it('should have display none when not checked', () => {
        const wrapper = shallow(<RadioButton />);
        expect(wrapper.find(StyledSpan)).toHaveStyleRule('display', 'none', {
            modifier: ':after',
        });
    });
});
