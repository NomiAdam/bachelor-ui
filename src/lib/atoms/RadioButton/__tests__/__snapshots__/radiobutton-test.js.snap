// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton should match the snapshot 1`] = `
<RadioButton__StyledContainer>
  <RadioButton__StyledLabel />
  <RadioButton__StyledInput
    type="radio"
  />
  <RadioButton__StyledSpan />
</RadioButton__StyledContainer>
`;

exports[`RadioButton should match the snapshot when checked 1`] = `
<RadioButton__StyledContainer>
  <RadioButton__StyledLabel />
  <RadioButton__StyledInput
    checked={true}
    type="radio"
  />
  <RadioButton__StyledSpan
    checked={true}
  />
</RadioButton__StyledContainer>
`;

exports[`RadioButton should match the snapshot with label 1`] = `
<RadioButton__StyledContainer>
  <RadioButton__StyledLabel>
    Mock
  </RadioButton__StyledLabel>
  <RadioButton__StyledInput
    type="radio"
  />
  <RadioButton__StyledSpan />
</RadioButton__StyledContainer>
`;
